tests/cedict-service/cedict.test.js

```diff
@@ -0,0 +1,279 @@
+    delete configuration.storage.stores.dictionary
+    expect(() => new Cedict(configuration)).toThrowError(Cedict.errMsgs.CONF_NO_DICT)
+  })
+
+  it('constructor: dictionary must have a primaryIndex subtree', () => {
+    let configuration = JSON.parse(JSON.stringify(CedictConfig)) // eslint-disable-line prefer-const
+    delete configuration.storage.stores.dictionary.primaryIndex
+    expect(() => new Cedict(configuration)).toThrowError(Cedict.errMsgs.CONF_NO_DICT_PRIMARY_IDX)
+  })
+
+  it('constructor: primaryIndex must have a keyPath', () => {
+    let configuration = JSON.parse(JSON.stringify(CedictConfig)) // eslint-disable-line prefer-const
+    delete configuration.storage.stores.dictionary.primaryIndex.keyPath
+    expect(() => new Cedict(configuration)).toThrowError(Cedict.errMsgs.CONF_NO_DICT_PRIMARY_IDX_KEY_PATH)
+  })
+
+  it('constructor: dictionary must have a volatileStorage tree', () => {
+    let configuration = JSON.parse(JSON.stringify(CedictConfig)) // eslint-disable-line prefer-const
+    delete configuration.storage.stores.dictionary.volatileStorage
+    expect(() => new Cedict(configuration)).toThrowError(Cedict.errMsgs.CONF_NO_DICT_VOLATILE)
+  })
+
+  it('constructor: volatileStorage must have an enabled key', () => {
+    let configuration = JSON.parse(JSON.stringify(CedictConfig)) // eslint-disable-line prefer-const
+    delete configuration.storage.stores.dictionary.volatileStorage.enabled
+    expect(() => new Cedict(configuration)).toThrowError(Cedict.errMsgs.CONF_NO_DICT_VOLATILE_ENABLED)
+  })
+
+  it('constructor: volatileStorage must have an indexed key', () => {
+    let configuration = JSON.parse(JSON.stringify(CedictConfig)) // eslint-disable-line prefer-const
+    delete configuration.storage.stores.dictionary.volatileStorage.indexed
+    expect(() => new Cedict(configuration)).toThrowError(Cedict.errMsgs.CONF_NO_DICT_VOLATILE_INDEXED)
+  })
+
+  it('constructor: dictionary must have a permanentStorage tree', () => {
+    let configuration = JSON.parse(JSON.stringify(CedictConfig)) // eslint-disable-line prefer-const
+    delete configuration.storage.stores.dictionary.permanentStorage
+    expect(() => new Cedict(configuration)).toThrowError(Cedict.errMsgs.CONF_NO_DICT_PERM)
+  })
+
+  it('constructor: permanentStorage must have an enabled key', () => {
+    let configuration = JSON.parse(JSON.stringify(CedictConfig)) // eslint-disable-line prefer-const
+    delete configuration.storage.stores.dictionary.permanentStorage.enabled
+    expect(() => new Cedict(configuration)).toThrowError(Cedict.errMsgs.CONF_NO_DICT_PERM_ENABLED)
+  })
+
+  it('constructor: permanentStorage must have an indexed key', () => {
+    let configuration = JSON.parse(JSON.stringify(CedictConfig)) // eslint-disable-line prefer-const
+    delete configuration.storage.stores.dictionary.permanentStorage.indexed
+    expect(() => new Cedict(configuration)).toThrowError(Cedict.errMsgs.CONF_NO_DICT_PERM_INDEXED)
+  })
```

Comment:
just wondering if these tests, and the underlying code, actually all belong in the permanent storage class rather than the cedict class -- are they specific to the cedict use of permanent storage or a feature of permanant storage itself?

Reply:
I think you are right that those should be part of CedictPermaentStorage and I shall move them there. However, we discussed making config a JSON and using JSON Schema to validate it, so maybe I'll synchronize with those changes. Would that be fine?

Reply:
yep, sounds good.